File: backend/app/schemas/onboarding.py
"""
Onboarding Pydantic Schemas

This module defines the Pydantic models for onboarding-related
API requests and responses.
"""

from pydantic import BaseModel, Field, ConfigDict
from uuid import UUID
from typing import Optional
from datetime import datetime


class OnboardingStatus(BaseModel):
    """
    Onboarding status response model
    
    Used by GET /api/onboarding/status endpoint to return the user's
    current onboarding progress and state.
    """
    onboarding_completed: bool = Field(..., description="Whether user has completed the onboarding flow")
    onboarding_step: Optional[str] = Field(None, description="Current step in onboarding flow (welcome, recommendations, complete)")
    onboarding_skipped: bool = Field(..., description="Whether user skipped the onboarding flow")
    tooltip_tour_completed: bool = Field(..., description="Whether user has completed the tooltip tour")
    should_show_onboarding: bool = Field(..., description="Whether the onboarding modal should be displayed to the user")

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "onboarding_completed": False,
                "onboarding_step": "welcome",
                "onboarding_skipped": False,
                "tooltip_tour_completed": False,
                "should_show_onboarding": True
            }
        }
    )


class UpdateOnboardingProgressRequest(BaseModel):
    """
    Request model for updating onboarding progress
    
    Used by POST /api/onboarding/progress endpoint.
    """
    step: str = Field(..., description="Onboarding step (welcome, recommendations, complete)")
    completed: bool = Field(..., description="Whether the step is completed")

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "step": "welcome",
                "completed": True
            }
        }
    )


class UserPreferences(BaseModel):
    """
    Complete user preferences model
    
    Represents the full user_preferences table record.
    """
    id: UUID = Field(..., description="Preference record UUID")
    user_id: UUID = Field(..., description="User UUID")
    onboarding_completed: bool = Field(default=False, description="Whether user has completed the onboarding flow")
    onboarding_step: Optional[str] = Field(None, description="Current step in onboarding flow")
    onboarding_skipped: bool = Field(default=False, description="Whether user skipped the onboarding flow")
    onboarding_started_at: Optional[datetime] = Field(None, description="Timestamp when user started onboarding")
    onboarding_completed_at: Optional[datetime] = Field(None, description="Timestamp when user completed onboarding")
    tooltip_tour_completed: bool = Field(default=False, description="Whether user has completed the tooltip tour")
    tooltip_tour_skipped: bool = Field(default=False, description="Whether user skipped the tooltip tour")
    preferred_language: str = Field(default='zh-TW', description="User preferred language code")
    created_at: datetime = Field(..., description="Record creation timestamp")
    updated_at: datetime = Field(..., description="Record last update timestamp")

    model_config = ConfigDict(
        from_attributes=True,
        json_schema_extra={
            "example": {
                "id": "123e4567-e89b-12d3-a456-426614174000",
                "user_id": "123e4567-e89b-12d3-a456-426614174001",
                "onboarding_completed": False,
                "onboarding_step": "welcome",
                "onboarding_skipped": False,
                "onboarding_started_at": "2024-01-01T00:00:00Z",
                "onboarding_completed_at": None,
                "tooltip_tour_completed": False,
                "tooltip_tour_skipped": False,
                "preferred_language": "zh-TW",
                "created_at": "2024-01-01T00:00:00Z",
                "updated_at": "2024-01-01T00:00:00Z"
            }
        }
    )

File: frontend/lib/api/onboarding.ts
import { apiClient } from './client';

/**
 * Onboarding API functions
 *
 * Provides methods for managing user onboarding flow.
 */

/**
 * Onboarding step type
 */
export type OnboardingStep =
  | 'welcome'
  | 'profile_setup'
  | 'feed_selection'
  | 'preferences'
  | 'tutorial'
  | 'complete';

/**
 * Onboarding status response
 */
export interface OnboardingStatus {
  is_completed: boolean;
  is_skipped: boolean;
  current_step?: OnboardingStep;
  completed_steps: OnboardingStep[];
  started_at?: string;
  completed_at?: string;
  skipped_at?: string;
}

/**
 * Update onboarding progress request
 */
export interface UpdateOnboardingProgressRequest {
  step: OnboardingStep;
  completed: boolean;
  metadata?: Record<string, any>;
}

/**
 * Get current onboarding status
 *
 * @returns Promise resolving to onboarding status
 */
export async function getOnboardingStatus(): Promise<OnboardingStatus> {
  return apiClient.get<OnboardingStatus>('/api/onboarding/status');
}

/**
 * Update onboarding progress
 *
 * @param step - Onboarding step to update
 * @param completed - Whether the step is completed
 * @param metadata - Optional metadata for the step
 * @returns Promise resolving to success response
 */
export async function updateOnboardingProgress(
  step: OnboardingStep,
  completed: boolean,
  metadata?: Record<string, any>
): Promise<{ message: string }> {
  return apiClient.post('/api/onboarding/progress', {
    step,
    completed,
    metadata,
  });
}

/**
 * Mark onboarding as completed
 *
 * @returns Promise resolving to success response
 */
export async function markOnboardingCompleted(): Promise<{ message: string }> {
  return apiClient.post('/api/onboarding/complete', {});
}

/**
 * Mark onboarding as skipped
 *
 * @returns Promise resolving to success response
 */
export async function markOnboardingSkipped(): Promise<{ message: string }> {
  return apiClient.post('/api/onboarding/skip', {});
}

/**
 * Reset onboarding status
 *
 * @returns Promise resolving to success response
 */
export async function resetOnboarding(): Promise<{ message: string }> {
  return apiClient.post('/api/onboarding/reset', {});
}
